File: packages/python/src/sah_sdk/zyins/reference/autocorrector.py
"""``Autocorrector`` adapter — typo correction on free-text input.

Mirror of ``packages/ts/src/zyins/reference/Autocorrector.ts`` and a
line-for-line semantic port of the bpp2.0 reference algorithm at
``src/sah-ui/Input/TextField/useAutocorrect.js``.

The adapter exists so consumers can:

  * inject a custom :class:`Autocorrector` (wholesale replacement),
  * decorate :class:`DefaultAutocorrector` via ``.clone(...)`` (token
    overlay), or
  * compose multiple correctors in a wrapper that itself implements the
    Protocol (chain).

The same three-pattern surface ships in every SDK language.

Example
-------

    >>> from sah_sdk.zyins.reference.autocorrector import DefaultAutocorrector
    >>> typo_map = {"ASTHM": "ASTHMA", "CHOLESTEROL": "HIGH CHOLESTEROL"}
    >>> ac = DefaultAutocorrector(typo_map=typo_map)
    >>> ac.correct("astm", mode="submit")
    'ASTM'
    >>> ac.correct("cholesterol", mode="submit")
    'HIGH CHOLESTEROL'
    >>> ac.correct("ASTHM", mode="keyup")  # mid-typing guard
    'ASTHM'
"""

from __future__ import annotations

import re
from collections.abc import Callable, Mapping
from dataclasses import dataclass, field, replace
from typing import Any, Literal, Protocol, runtime_checkable

# Whitespace-splitter mirrors the JS algorithm's ``/\s+/`` regex.
_WHITESPACE_RE = re.compile(r"\s+")


def _contains_phrase(words: list[str], phrase: str) -> bool:
    """Return True when ``phrase`` appears as a consecutive token run.

    Mirrors the locked TS ``containsPhrase``: the phrase is whitespace-tokenized
    and matched as a contiguous subsequence of ``words`` — never as a substring
    inside a single token.
    """
    phrase_words = [w for w in _WHITESPACE_RE.split(phrase) if w]
    if not phrase_words or len(phrase_words) > len(words):
        return False
    last_start = len(words) - len(phrase_words)
    for start in range(last_start + 1):
        if words[start : start + len(phrase_words)] == phrase_words:
            return True
    return False


AutocorrectMode = Literal["keyup", "submit"]


@dataclass(frozen=True, slots=True)
class AutocorrectAppliedEvent:
    """Emitted to ``on_applied`` when a correction is applied.

    ``original`` is the verbatim text passed to :meth:`Autocorrector.correct`.
    ``corrected`` is the post-correction output. ``mode`` is the call mode.
    ``window`` and ``correction`` identify the specific typo replacement
    that triggered this event.
    """

    original: str
    corrected: str
    mode: AutocorrectMode
    window: str = ""
    correction: str = ""


@runtime_checkable
class Autocorrector(Protocol):
    """Protocol — typo-correct free-text input.

    Implementations MUST be side-effect-free with respect to the
    input string; they MUST preserve trailing whitespace; they MUST
    return the input unchanged when no correction applies.

    ``mode``:
      * ``"keyup"`` — the user is mid-typing. Skip corrections whose
        replacement contains the input as a substring AND is longer
        than the input (prevents "ASTHM" → "ASTHMA" while still typing).
      * ``"submit"`` — the input is final. Skip corrections whose
        replacement is already present as a substring of the input
        (prevents "HIGH CHOLESTEROL" → "HIGH HIGH CHOLESTEROL").
    """

    def correct(self, text: str, *, mode: AutocorrectMode) -> str:
        """Return ``text`` with typo corrections applied per ``mode``."""
        ...


@dataclass(frozen=True, slots=True)
class _AutocorrectorConfig:
    """Frozen config bag for :class:`DefaultAutocorrector`.

    Kept private so callers reach for :meth:`DefaultAutocorrector.clone`
    instead of mutating a config object directly.
    """

    typo_map: Mapping[str, str]
    version_tag: str | None = None
    on_applied: Callable[[AutocorrectAppliedEvent], None] | None = field(
        default=None, compare=False, repr=False
    )


class DefaultAutocorrector:
    """Reference implementation — n-gram sliding-window over a typo map.

    Algorithm (line-for-line semantic port of bpp2.0
    ``useAutocorrect.js``):

      1. Trim trailing whitespace, remember whether one was present.
      2. Uppercase the input, split on whitespace.
      3. For ``num_words`` in ``0..len(words)`` and each starting index
         ``i``, consider the contiguous n-gram ``words[i:i+num_words+1]``
         joined by a single space.
      4. Look it up in ``typo_map`` (keys are pre-uppercased by the
         caller). On a hit, apply the mode-specific guard:
         - ``keyup``: skip if ``correction.upper().includes(window)`` AND
           ``len(correction) > len(window)``.
         - ``submit``: skip if ``upper_input.includes(correction)``.
      5. If the guard passes, replace the slot at ``i`` with the
         correction; mark every position in the n-gram as processed;
         skip ahead past the consumed words.
      6. Fill un-processed slots with their original (uppercased) words.
      7. Re-join with single spaces; re-append the saved trailing
         whitespace.

    The implementation is single-threaded and allocates an O(n) list
    per call; no shared mutable state outside the optional
    ``on_applied`` callback fires once per applied correction.
    """

    __slots__ = ("_config",)

    _config: _AutocorrectorConfig

    def __init__(
        self,
        *,
        typo_map: Mapping[str, str],
        version_tag: str | None = None,
        on_applied: Callable[[AutocorrectAppliedEvent], None] | None = None,
    ) -> None:
        """Construct from a pre-uppercased typo map.

        :param typo_map: ``{"HBP": "HIGH BLOOD PRESSURE"}`` — keys and
            values are uppercase, single-space separated. The map is
            captured by reference; callers MUST NOT mutate it after
            construction.
        :param version_tag: Optional opaque tag identifying the
            underlying dataset version. Surfaced as :attr:`version_tag`
            for telemetry; never inspected by the algorithm.
        :param on_applied: Optional callback fired once per correction
            applied. Exceptions raised by the callback propagate to the
            caller of :meth:`correct`.
        """
        self._config = _AutocorrectorConfig(
            typo_map=typo_map,
            version_tag=version_tag,
            on_applied=on_applied,
        )

    @property
    def version_tag(self) -> str | None:
        """Opaque dataset-version tag, or ``None`` when unspecified."""
        return self._config.version_tag

    def clone(self, **overrides: Any) -> DefaultAutocorrector:
        """Return a fresh instance with selected fields overridden.

        Accepts the same keyword set as :meth:`__init__`. Fields not
        present in ``overrides`` are carried over from the current
        instance.

        >>> ac = DefaultAutocorrector(typo_map={"HBP": "HIGH BLOOD PRESSURE"})
        >>> child = ac.clone(version_tag="2026.05.29")
        >>> child.version_tag
        '2026.05.29'
        """
        new_config = replace(self._config, **overrides)
        return DefaultAutocorrector(
            typo_map=new_config.typo_map,
            version_tag=new_config.version_tag,
            on_applied=new_config.on_applied,
        )

    def correct(self, text: str, *, mode: AutocorrectMode) -> str:
        """Apply typo corrections per the locked algorithm.

        Exceptions raised by ``on_applied`` callbacks propagate to the caller.
        """
        if not text:
            return text

        trailing = " " if text.endswith(" ") else ""
        upper_text = text.upper()
        words = _WHITESPACE_RE.split(upper_text.strip()) if upper_text.strip() else []
        if not words:
            return text
        # An empty typo map still normalizes: uppercase + whitespace-collapse +
        # trailing-space preservation, matching the locked TS adapter (which
        # returns ``words.join(' ') + trailing`` on a zero-size map rather than
        # the raw input).
        if not self._config.typo_map:
            return " ".join(words) + trailing

        slots: list[str | None] = [None] * len(words)
        consumed: set[int] = set()
        typo_map = self._config.typo_map
        applied: list[tuple[str, str]] = []

        # Mirror the JS double-loop. The reference processes larger phrases
        # first (descending window size) so a multi-word correction wins over
        # a competing single-word one, and considers only exact-size n-grams —
        # a clamped tail slice shorter than the window is skipped and handled
        # by a later, smaller pass. The windowSize=0 pass is the size-1 sweep
        # that reaches every single word, so descending order still covers the
        # tail. See the locked TS DefaultAutocorrector.
        for window_size in range(len(words) - 1, -1, -1):
            i = 0
            while i < len(words):
                window_words = words[i : i + window_size + 1]
                if len(window_words) != window_size + 1:
                    i += 1
                    continue
                window_len = len(window_words)
                window = " ".join(window_words)
                correction = typo_map.get(window)
                if correction is None:
                    i += 1
                    continue

                if mode == "keyup":
                    # Keyup-mode guard: skip if correction is a strict
                    # superstring of the window (mid-typing protection).
                    correction_upper = correction.upper()
                    if window in correction_upper and len(correction) > len(window):
                        i += 1
                        continue
                else:
                    # Submit-mode guard: anti-duplication — skip only when the
                    # correction already appears as a consecutive whitespace-
                    # token phrase in the input, NOT when it is merely a
                    # substring inside a single token. Mirrors the locked TS
                    # ``containsPhrase``.
                    if _contains_phrase(words, correction.upper()):
                        i += 1
                        continue

                if slots[i] is not None:
                    i += 1
                    continue
                if any((i + n) in consumed for n in range(window_len)):
                    i += 1
                    continue

                slots[i] = correction
                for n in range(window_len):
                    if i + n < len(words):
                        consumed.add(i + n)
                        if n > 0:
                            slots[i + n] = None
                applied.append((window, correction))
                i += 1

        out_words: list[str] = []
        for index, word in enumerate(words):
            if slots[index] is not None:
                out_words.append(slots[index] or "")
            elif index not in consumed:
                out_words.append(word)
        result = " ".join(out_words)
        # Collapse internal repeats the way the JS algorithm does via
        # ``.split(/\s+/).join(' ')`` after the replacement pass.
        result = " ".join(_WHITESPACE_RE.split(result)) + trailing

        if applied and self._config.on_applied is not None:
            for window, correction in applied:
                self._config.on_applied(
                    AutocorrectAppliedEvent(
                        original=text,
                        corrected=result,
                        mode=mode,
                        window=window,
                        correction=correction,
                    )
                )
        return result


def create(
    *,
    typo_map: Mapping[str, str],
    version_tag: str | None = None,
    on_applied: Callable[[AutocorrectAppliedEvent], None] | None = None,
) -> DefaultAutocorrector:
    """Factory mirroring ``isa.autocorrector.create({ typoMap })`` (TS).

    The kernel-style entry point. Equivalent to instantiating
    :class:`DefaultAutocorrector` directly; documented separately so the
    surface name matches the locked cross-language SDK syntax.
    """
    return DefaultAutocorrector(
        typo_map=typo_map, version_tag=version_tag, on_applied=on_applied
    )


__all__ = [
    "AutocorrectAppliedEvent",
    "AutocorrectMode",
    "Autocorrector",
    "DefaultAutocorrector",
    "create",
]

File: packages/ts/src/zyins/reference/Autocorrector.ts
/**
 * `Autocorrector` — free-text typo correction for picker UIs.
 *
 * The interface accepts free text and returns the same text with known
 * typos rewritten. Mode governs typing-state heuristics:
 *   - `'keyup'`: live correction during typing. Skip a correction whose
 *     RHS contains the typed token AND is longer — prevents `ASTHM` from
 *     prematurely becoming `ASTHMA` while still typing.
 *   - `'submit'`: post-commit correction. Skip a correction whose RHS is
 *     already present in the surrounding text — prevents
 *     `HIGH CHOLESTEROL` from becoming `HIGH HIGH CHOLESTEROL`.
 *
 * Algorithm parity: the default implementation is a direct port of
 * `bpp2.0/src/sah-ui/Input/TextField/useAutocorrect.js` (the n-gram
 * sliding window over uppercase tokens). Semantically identical so the
 * SDK can replace the bpp2.0 hook drop-in.
 *
 * @example
 * ```ts
 * // Domain-bound (typoMap pre-loaded from zyins datasets):
 * const corrected = isa.zyins.autocorrector.correct('hyprtension', { mode: 'keyup' });
 * // → 'HYPERTENSION'
 *
 * // Custom typo map (top-level kernel):
 * const corrector = isa.autocorrector.create({
 *   typoMap: new Map([['HYPRTENSION', 'HYPERTENSION']]),
 * });
 * corrector.correct('hyprtension', { mode: 'submit' });
 * ```
 *
 * Substitution: replace via {@link DefaultAutocorrector.clone} or pass
 * a custom `Autocorrector` to {@link Isa.withKeycode}.
 */
export interface Autocorrector {
  /**
   * Apply typo corrections to free-text input.
   *
   * Never throws. Empty input returns empty. Trailing whitespace on the
   * input is preserved on the output so cursor positioning in a text
   * field remains stable across edits.
   */
  correct(text: string, options: AutocorrectOptions): string;
}

/** Mode + per-call options for {@link Autocorrector.correct}. */
export interface AutocorrectOptions {
  /**
   * Typing-state mode. `'keyup'` for live correction during typing;
   * `'submit'` for post-commit (blur, enter, form submit) correction.
   */
  readonly mode: 'keyup' | 'submit';
}

/** Event emitted by {@link DefaultAutocorrector.onApplied}. */
export interface AutocorrectAppliedEvent {
  /** Pre-correction input, verbatim. */
  readonly input: string;
  /** Post-correction output. */
  readonly output: string;
  /** Mode the consumer passed. */
  readonly mode: 'keyup' | 'submit';
}

/** Constructor options for {@link DefaultAutocorrector}. */
export interface DefaultAutocorrectorOptions {
  /**
   * Map of UPPERCASE input → UPPERCASE correction. Keys MUST be the
   * normalized form (uppercase tokens, joined by a single space).
   * Conventionally sourced from
   * `isa.zyins.datasets.spellingCorrections.items`.
   */
  readonly typoMap: ReadonlyMap<string, string>;
  /**
   * Optional version stamp for the typo map. Surface via {@link
   * DefaultAutocorrector.versionTag} so consumers can detect a stale
   * binding and refresh it.
   */
  readonly versionTag?: string;
  /**
   * Optional sink invoked whenever a correction actually changes the
   * text. Useful for analytics ("user typed X, we corrected to Y").
   * Fires AFTER the text is computed and BEFORE `correct()` returns.
   */
  readonly onApplied?: (event: AutocorrectAppliedEvent) => void;
}

/**
 * Default autocorrector. Implements the n-gram sliding-window algorithm
 * from the bpp2.0 picker hook.
 *
 * Single-threaded, synchronous, dependency-free. Safe to share across
 * concurrent calls — the instance holds no mutable per-call state.
 *
 * @example
 * ```ts
 * const corrector = new DefaultAutocorrector({
 *   typoMap: new Map([['HYPRTENSION', 'HYPERTENSION']]),
 *   versionTag: '2026.05.29',
 *   onApplied: (e) => analytics.track('autocorrect', e),
 * });
 *
 * corrector.correct('hyprtension', { mode: 'submit' });
 * // → 'HYPERTENSION'
 *
 * // Replace the typo map without losing the listener:
 * const next = corrector.clone({ typoMap: freshMap, versionTag: '2026.06.01' });
 * ```
 */
export class DefaultAutocorrector implements Autocorrector {
  private readonly typoMap: ReadonlyMap<string, string>;
  private readonly _versionTag: string | undefined;
  private readonly _onApplied: ((event: AutocorrectAppliedEvent) => void) | undefined;

  constructor(opts: DefaultAutocorrectorOptions) {
    this.typoMap = opts.typoMap;
    this._versionTag = opts.versionTag;
    this._onApplied = opts.onApplied;
  }

  /** Opaque tag tracking the version of the typo map this corrector binds. */
  get versionTag(): string | undefined {
    return this._versionTag;
  }

  /**
   * Apply typo corrections.
   *
   * Mirrors the bpp2.0 algorithm:
   *   1. Uppercase the input; tokenize on whitespace.
   *   2. For window size 1..wordCount, slide every contiguous n-gram.
   *   3. Look up the n-gram in `typoMap`. Apply guard for the mode.
   *   4. Replace the matched span; mark positions consumed.
   *   5. Fill non-matched positions with the original (uppercased) word.
   *   6. Reassemble with single-space separator; preserve trailing space.
   */
  correct(text: string, options: AutocorrectOptions): string {
    if (!text) return text;

    const trailingWhitespace = text.endsWith(' ') ? ' ' : '';
    const upper = text.toUpperCase();
    // Match PHP's PREG_SPLIT_NO_EMPTY / Python's str.split() — drop empty
    // tokens that .split(/\s+/) emits at string boundaries. Otherwise a
    // trailing space produces an empty trailing token that survives all
    // the way to the join and concatenates with `trailingWhitespace` to
    // emit a double space (cross-language SDK contract requires single).
    const words = upper.split(/\s+/).filter((w) => w !== '');
    if (words.length === 0) return text;
    if (this.typoMap.size === 0) return words.join(' ') + trailingWhitespace;
    const result: (string | undefined)[] = new Array<string | undefined>(words.length);
    const consumed = new Set<number>();
    let changed = false;

    // Process larger phrases first while keeping each pass to its exact
    // n-gram size; shorter tail slices are handled by later passes.
    for (let windowSize = words.length - 1; windowSize >= 0; windowSize--) {
      for (let i = 0; i < words.length; i++) {
        const ngram = words.slice(i, i + windowSize + 1);
        if (ngram.length !== windowSize + 1) continue;
        const key = ngram.join(' ');

        const correction = this.typoMap.get(key);
        if (correction === undefined) continue;

        const correctionUpper = correction.toUpperCase();
        const shouldCorrect =
          options.mode === 'keyup'
            ? !(correctionUpper.includes(key) && correctionUpper.length > key.length)
            : !containsPhrase(words, correctionUpper);

        if (!shouldCorrect) continue;
        if (result[i] !== undefined) continue;
        if (windowOverlapsConsumed(consumed, i, ngram.length)) continue;

        changed = true;
        result[i] = correction;
        const windowLen = ngram.length;
        for (let n = 0; n < windowLen; n++) {
          const consumedIndex = i + n;
          if (consumedIndex >= words.length) continue;
          consumed.add(consumedIndex);
          if (n > 0) result[consumedIndex] = undefined;
        }
      }
    }

    for (let i = 0; i < words.length; i++) {
      if (result[i] === undefined && !consumed.has(i)) {
        result[i] = words[i];
      }
    }

    // Filter out the undefineds (positions that were absorbed into a
    // multi-word correction at a lower index) before joining.
    const joined =
      result.filter((x): x is string => x !== undefined).join(' ') + trailingWhitespace;

    if (changed && this._onApplied !== undefined) {
      this._onApplied({ input: text, output: joined, mode: options.mode });
    }

    return joined;
  }

  /**
   * Return a new corrector with selected fields overridden. Useful for
   * swapping the typo map after a fresh dataset bundle arrives without
   * losing the `onApplied` sink.
   *
   * @example
   * ```ts
   * const fresh = corrector.clone({ typoMap: nextMap, versionTag: '2026.06.01' });
   * ```
   */
  clone(overrides: Partial<DefaultAutocorrectorOptions> = {}): DefaultAutocorrector {
    const nextVersionTag = overrides.versionTag ?? this._versionTag;
    const nextOnApplied = overrides.onApplied ?? this._onApplied;
    return new DefaultAutocorrector({
      typoMap: overrides.typoMap ?? this.typoMap,
      ...(nextVersionTag !== undefined && { versionTag: nextVersionTag }),
      ...(nextOnApplied !== undefined && { onApplied: nextOnApplied }),
    });
  }
}

function containsPhrase(words: readonly string[], phrase: string): boolean {
  const phraseWords = phrase.split(/\s+/).filter((w) => w !== '');
  if (phraseWords.length === 0 || phraseWords.length > words.length) return false;
  for (let start = 0; start <= words.length - phraseWords.length; start++) {
    const matches = phraseWords.every((word, offset) => words[start + offset] === word);
    if (matches) return true;
  }
  return false;
}

function windowOverlapsConsumed(
  consumed: ReadonlySet<number>,
  start: number,
  length: number,
): boolean {
  for (let offset = 0; offset < length; offset++) {
    if (consumed.has(start + offset)) return true;
  }
  return false;
}
